### PR TITLE
Add commands for outbox ls and clear.

### PR DESCRIPTION
**Pull request by anorth**
Also adds a handy SilentWriter for writing command output without too much error handling.

Related to #2165 and #2146.

### CODE REVIEW

File: commands/utils.go [hunk 1]
```diff
@@ -7,13 +7,80 @@ import (
...
+	if sw.err == nil {
+		_, sw.err = io.WriteString(sw.w, str)
+	}
+	return sw.err == nil
+}
+
+// Print writes with fmt.Print
+func (sw *silentWriter) Print(a ...interface{}) bool {
+	if sw.err == nil {
+		_, sw.err = fmt.Print(a...)
```
Comment (travisperson) [inline commands/utils.go]:
```suggestion
		_, sw.err = fmt.Fprint(sw.w, a...)
```

File: commands/utils.go [hunk 2]
```diff
@@ -7,13 +7,80 @@ import (
...
+	}
+	return sw.err == nil
+}
+
+// Print writes with fmt.Println
+func (sw *silentWriter) Println(a ...interface{}) bool {
+	if sw.err == nil {
+		_, sw.err = fmt.Println(a...)
```
Comment (travisperson) [inline commands/utils.go]:
```suggestion
		_, sw.err = fmt.Fprintln(sw.w, a...)
```

File: commands/utils.go [hunk 3]
```diff
@@ -7,13 +7,80 @@ import (
...
+	}
+	return sw.err == nil
+}
+
+// Print writes with fmt.Printf
+func (sw *silentWriter) Printf(format string, a ...interface{}) bool {
+	if sw.err == nil {
+		_, sw.err = fmt.Printf(format, a...)
```
Comment (travisperson) [inline commands/utils.go]:
```suggestion
		_, sw.err = fmt.Fprintf(sw.w, format, a...)
```

File: commands/outbox.go [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+	"github.com/filecoin-project/go-filecoin/address"
+	"github.com/filecoin-project/go-filecoin/core"
+)
+
+var outboxCmd = &cmds.Command{
+	Helptext: cmdkit.HelpText{
+		Tagline: "View and manipulate the outbound message queue",
+	},
+	Subcommands: map[string]*cmds.Command{
+		"ls":    outboxLsCmd,
```
Comment (phritz) [inline commands/outbox.go]: i know its pedantic but lexicographic order as these things grow is always helpful

File: commands/outbox.go [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+		cmds.Text: cmds.MakeTypedEncoder(func(req *cmds.Request, w io.Writer, queue *OutboxLsResult) error {
+			sw := NewSilentWriter(w)
+			sw.Println("From:", queue.Address.String())
+			for _, qm := range queue.Messages {
+				msg := qm.Msg
+				cid, err := msg.Cid()
+				if err != nil {
+					return err
+				}
+				sw.Printf("CID: %s, nonce: %d, gas limit: %d, height: %d\n", cid.String(), msg.Nonce, msg.GasLimit, qm.Stamp)
```
Comment (phritz) [inline commands/outbox.go]: maybe we should just print msg.string() and then the extra fields?

File: commands/outbox_daemon_test.go [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+		d.RunSuccess("outbox", "clear", fixtures.TestAddresses[1])
+		out := d.RunSuccess("outbox", "ls").ReadStdout()
+		assert.Contains(out, fixtures.TestAddresses[0])
+		assert.NotContains(out, fixtures.TestAddresses[1]) // Cleared
+		assert.Contains(out, c1)
+		assert.Contains(out, c2)
+		assert.NotContains(out, c3) // cleared
+
+		// #nofilter
+		d.RunSuccess("outbox", "clear")
```
Comment (phritz) [inline commands/outbox_daemon_test.go]: might want to repopulate with the messages above before clearing so we know that it clears multiple addresses. also probably worth calling clear on a the empty queue afterwards so we can be sure it doesn't error.

File: commands/utils.go [hunk 7]
```diff
@@ -7,13 +7,80 @@ import (
...
+
+// Error clears and returns any error encountered while writing.
+// Subsequent writes will attempt to write to the underlying writer again.
+func (sw *silentWriter) ClearError() error {
+	err := sw.err
+	sw.err = nil
+	return err
+}
+
+// Write writes with io.Writer.Write
```
Comment (phritz) [inline commands/utils.go]: maybe: "and returns true if there was no error" here and below

File: commands/utils.go (hunk 1 above)
Comment (phritz) [reply commands/utils.go]: indeed

File: commands/utils.go [hunk 8]
```diff
@@ -7,13 +7,80 @@ import (
...
+// silentWriter writes to a stream, stopping after the first error and discarding output until
+// the error is cleared.
+// No printing methods return an error (to avoid warnings about ignoring it), but they do return
+// a boolean indicating whether an error is waiting to be cleared.
+// Example usage:
+//   sw := NewSilentWriter(w)
+//   sw.Println("A line")
+//   sw.Println("Another line")
+//   return sw.Error()
+type silentWriter struct {
```
Comment (phritz) [inline commands/utils.go]: love it, hey @acruikshank check it out

File: core/message_queue.go [hunk 9]
```diff
@@ -121,6 +118,20 @@ func (mq *MessageQueue) LargestNonce(sender address.Address) (largest uint64, fo
 	return 0, false
 }
 
+// Queues returns the addresses associated with each non-empty queue.
```
Comment (phritz) [inline core/message_queue.go]: maybe note that there is no guarantee the addresses are in any particular or consistent order (bc the map iteration).

File: core/message_queue.go [hunk 10]
```diff
@@ -83,11 +83,8 @@ func (mq *MessageQueue) Clear(sender address.Address) bool {
 	defer mq.lk.Unlock()
 
 	q := mq.queues[sender]
-	if len(q) > 0 {
-		mq.queues[sender] = []*QueuedMessage{}
-		return true
-	}
-	return false
+	delete(mq.queues, sender)
+	return len(q) > 0
```
Comment (phritz) [inline core/message_queue.go]: maybe comment return value above

File: plumbing/api.go [hunk 11]
```diff
@@ -162,6 +165,21 @@ func (api *API) DealPut(storageDeal *storagedeal.Deal) error {
 	return api.storagedeals.Put(storageDeal)
 }
 
+// OutboxQueues lists addresses with non-empty outbox queues
```
Comment (phritz) [inline plumbing/api.go]: maybe note the nondeterminism of the order here to save someone a little heartache down the line

File: commands/utils.go (hunk 1 above)
Comment (anorth) [reply commands/utils.go]: Eek, thanks. Good catch!

**Pull request merged**